source/s3_client.c

```diff
@@ -1528,9 +1528,49 @@ static void s_s3_client_process_work_default(struct aws_s3_client *client) {
             AWS_LS_S3_CLIENT, "id=%p Updating connections, assigning requests where possible.", (void *)client);
         aws_s3_client_update_connections_threaded(client);
     }
+    /*******************/
+    /* Step 4: Cleanup Endpoints if Required. */
+    /*******************/
+
+    /* BEGIN CRITICAL SECTION */
+    aws_s3_client_lock_synced_data(client);
+
+    if (client->synced_data.process_endpoint_lifecycle_changes) {
+        client->synced_data.process_endpoint_lifecycle_changes = false;
+        for (struct aws_hash_iter iter = aws_hash_iter_begin(&client->synced_data.endpoints);
+             !aws_hash_iter_done(&iter);
+             aws_hash_iter_next(&iter)) {
+            struct aws_s3_endpoint *endpoint = (struct aws_s3_endpoint *)iter.element.value;
```

Comment:
An alternative approach that I considered was maintaining two linked lists of endpoints: `pending_cleanup_task` and `cleanup_task_scheduled`, which would allow us to only loop through endpoints that have the required state. However, considering that by default an AWS account has a limit of 100 buckets, I thought it was not worth the complexity.